### ISSUE TITLE
Nomad Clients cannot discover servers with Consul 1.3.0

### ISSUE THREAD

**Issue by lawliet89** [hashicorp/nomad#4804]
### Nomad version
Nomad v0.8.6 (ab54ebcfcde062e9482558b7c052702d4cb8aa1b+CHANGES)

### Operating system and Environment details

Ubuntu 16.04.

Configuration file:

```hcl
acl {
  enabled = true
}

datacenter = "ap-southeast-1c"
name       = "xxxxxx"
region     = "ap-southeast-1"
bind_addr  = "0.0.0.0"

advertise {
  http = "10.158.242.241"
  rpc  = "10.158.242.241"
  serf = "10.158.242.241"
}

client {
  enabled = true
}

consul {
  address = "127.0.0.1:8500"
}

```

### Issue

Using the exact configuration on the same machine, once I upgrade Consul from 1.2.3 to Consul 1.3.0, Nomad clients can no longer discover Nomad servers. If I downgrade Consul back to 1.2.3, the problem goes away.

The local Consul agent works and does return the list of Nomad servers:

```
$ curl -sS http://127.0.0.1:8500/v1/catalog/service/nomad | jq 'length'
15
```

### Nomad Client logs (if appropriate)

```
    2018/10/18 06:05:11.526186 [INFO] client: using state directory /opt/nomad/data/client
    2018/10/18 06:05:11.530831 [INFO] client: using alloc directory /opt/nomad/data/alloc
    2018/10/18 06:05:11.533112 [INFO] fingerprint.cgroups: cgroups are available
    2018/10/18 06:05:11.534599 [INFO] fingerprint.consul: consul agent is available
    2018/10/18 06:05:11.542545 [WARN] fingerprint.network: Unable to parse Speed in output of '/sbin/ethtool ens5'
    2018/10/18 06:05:11.611959 [INFO] fingerprint.vault: Vault is available
    2018/10/18 06:05:11.623234 [INFO] client: Node ID "227317f0-8038-a81d-c30f-ddeb5dda496d"
    2018/10/18 06:05:11.623409 [WARN] manager: No servers available
    2018/10/18 06:05:11.623554 [WARN] manager: No servers available
    2018/10/18 06:05:11.625411 [ERR] client.consul: error discovering nomad servers: no Nomad Servers advertising service "nomad" in Consul datacenters: ["ap-southeast-1"]
    2018/10/18 06:05:13.611716 [ERR] http: Request /v1/agent/health?type=client, error: {"client":{"ok":false,"message":"no known servers"}}
    2018/10/18 06:05:23.612704 [ERR] http: Request /v1/agent/health?type=client, error: {"client":{"ok":false,"message":"no known servers"}}
    2018/10/18 06:05:33.613480 [ERR] http: Request /v1/agent/health?type=client, error: {"client":{"ok":false,"message":"no known servers"}}
    2018/10/18 06:05:39.551829 [WARN] manager: No servers available
    2018/10/18 06:05:39.554980 [ERR] client.consul: error discovering nomad servers: no Nomad Servers advertising service "nomad" in Consul datacenters: ["ap-southeast-1"]
    2018/10/18 06:05:43.614248 [ERR] http: Request /v1/agent/health?type=client, error: {"client":{"ok":false,"message":"no known servers"}}
    2018/10/18 06:05:53.615119 [ERR] http: Request /v1/agent/health?type=client, error: {"client":{"ok":false,"message":"no known servers"}}
    2018/10/18 06:06:02.977477 [WARN] manager: No servers available
    2018/10/18 06:06:03.284782 [WARN] manager: No servers available
    2018/10/18 06:06:03.287338 [ERR] client.consul: error discovering nomad servers: no Nomad Servers advertising service "nomad" in Consul datacenters: ["ap-southeast-1"]
    2018/10/18 06:06:03.615838 [ERR] http: Request /v1/agent/health?type=client, error: {"client":{"ok":false,"message":"no known servers"}}
    2018/10/18 06:06:13.616613 [ERR] http: Request /v1/agent/health?type=client, error: {"client":{"ok":false,"message":"no known servers"}}
    2018/10/18 06:06:23.617403 [ERR] http: Request /v1/agent/health?type=client, error: {"client":{"ok":false,"message":"no known servers"}}
    2018/10/18 06:06:28.495440 [WARN] manager: No servers available
    2018/10/18 06:06:28.497697 [ERR] client.consul: error discovering nomad servers: no Nomad Servers advertising service "nomad" in Consul datacenters: ["ap-southeast-1"]
    2018/10/18 06:06:33.618125 [ERR] http: Request /v1/agent/health?type=client, error: {"client":{"ok":false,"message":"no known servers"}}
    2018/10/18 06:06:36.144176 [WARN] manager: No servers available
    2018/10/18 06:06:43.618924 [ERR] http: Request /v1/agent/health?type=client, error: {"client":{"ok":false,"message":"no known servers"}}
    2018/10/18 06:06:48.994052 [WARN] manager: No servers available
    2018/10/18 06:06:48.996348 [ERR] client.consul: error discovering nomad servers: no Nomad Servers advertising service "nomad" in Consul datacenters: ["ap-southeast-1"]
    2018/10/18 06:06:53.619722 [ERR] http: Request /v1/agent/health?type=client, error: {"client":{"ok":false,"message":"no known servers"}}
    2018/10/18 06:07:03.620501 [ERR] http: Request /v1/agent/health?type=client, error: {"client":{"ok":false,"message":"no known servers"}}
    2018/10/18 06:07:13.621303 [ERR] http: Request /v1/agent/health?type=client, error: {"client":{"ok":false,"message":"no known servers"}}
    2018/10/18 06:07:18.682098 [WARN] manager: No servers available
    2018/10/18 06:07:18.685135 [ERR] client.consul: error discovering nomad servers: no Nomad Servers advertising service "nomad" in Consul datacenters: ["ap-southeast-1"]
    2018/10/18 06:07:23.622063 [ERR] http: Request /v1/agent/health?type=client, error: {"client":{"ok":false,"message":"no known servers"}}
    2018/10/18 06:07:33.622874 [ERR] http: Request /v1/agent/health?type=client, error: {"client":{"ok":false,"message":"no known servers"}}
    2018/10/18 06:07:35.701971 [WARN] manager: No servers available
    2018/10/18 06:07:43.623670 [ERR] http: Request /v1/agent/health?type=client, error: {"client":{"ok":false,"message":"no known servers"}}
    2018/10/18 06:07:47.675848 [WARN] manager: No servers available
    2018/10/18 06:07:47.678731 [ERR] client.consul: error discovering nomad servers: no Nomad Servers advertising service "nomad" in Consul datacenters: ["ap-southeast-1"]
    2018/10/18 06:07:53.624409 [ERR] http: Request /v1/agent/health?type=client, error: {"client":{"ok":false,"message":"no known servers"}}
    2018/10/18 06:08:03.625648 [ERR] http: Request /v1/agent/health?type=client, error: {"client":{"ok":false,"message":"no known servers"}}
    2018/10/18 06:08:09.804232 [WARN] manager: No servers available
    2018/10/18 06:08:09.807094 [ERR] client.consul: error discovering nomad servers: no Nomad Servers advertising service "nomad" in Consul datacenters: ["ap-southeast-1"]
    2018/10/18 06:08:13.626417 [ERR] http: Request /v1/agent/health?type=client, error: {"client":{"ok":false,"message":"no known servers"}}
    2018/10/18 06:08:23.627193 [ERR] http: Request /v1/agent/health?type=client, error: {"client":{"ok":false,"message":"no known servers"}}
    2018/10/18 06:08:32.249604 [WARN] manager: No servers available
    2018/10/18 06:08:33.627957 [ERR] http: Request /v1/agent/health?type=client, error: {"client":{"ok":false,"message":"no known servers"}}
    2018/10/18 06:08:39.076800 [WARN] manager: No servers available
    2018/10/18 06:08:39.079671 [ERR] client.consul: error discovering nomad servers: no Nomad Servers advertising service "nomad" in Consul datacenters: ["ap-southeast-1"]
    2018/10/18 06:08:43.628758 [ERR] http: Request /v1/agent/health?type=client, error: {"client":{"ok":false,"message":"no known servers"}}
    2018/10/18 06:08:53.629546 [ERR] http: Request /v1/agent/health?type=client, error: {"client":{"ok":false,"message":"no known servers"}}
    2018/10/18 06:08:59.698945 [WARN] manager: No servers available
    2018/10/18 06:08:59.701875 [ERR] client.consul: error discovering nomad servers: no Nomad Servers advertising service "nomad" in Consul datacenters: ["ap-southeast-1"]
    2018/10/18 06:09:03.630377 [ERR] http: Request /v1/agent/health?type=client, error: {"client":{"ok":false,"message":"no known servers"}}
    2018/10/18 06:09:13.002083 [WARN] manager: No servers available
    2018/10/18 06:09:13.631226 [ERR] http: Request /v1/agent/health?type=client, error: {"client":{"ok":false,"message":"no known servers"}}
    2018/10/18 06:09:20.301762 [WARN] manager: No servers available
    2018/10/18 06:09:20.304562 [ERR] client.consul: error discovering nomad servers: no Nomad Servers advertising service "nomad" in Consul datacenters: ["ap-southeast-1"]
    2018/10/18 06:09:23.631994 [ERR] http: Request /v1/agent/health?type=client, error: {"client":{"ok":false,"message":"no known servers"}}
    2018/10/18 06:09:33.633413 [ERR] http: Request /v1/agent/health?type=client, error: {"client":{"ok":false,"message":"no known servers"}}
    2018/10/18 06:09:42.801710 [WARN] manager: No servers available
    2018/10/18 06:09:42.803198 [ERR] client.consul: error discovering nomad servers: no Nomad Servers advertising service "nomad" in Consul datacenters: ["ap-southeast-1"]
    2018/10/18 06:09:43.634203 [ERR] http: Request /v1/agent/health?type=client, error: {"client":{"ok":false,"message":"no known servers"}}
    2018/10/18 06:09:53.634981 [ERR] http: Request /v1/agent/health?type=client, error: {"client":{"ok":false,"message":"no known servers"}}
    2018/10/18 06:10:03.635708 [ERR] http: Request /v1/agent/health?type=client, error: {"client":{"ok":false,"message":"no known servers"}}
    2018/10/18 06:10:06.519896 [WARN] manager: No servers available
    2018/10/18 06:10:06.521417 [ERR] client.consul: error discovering nomad servers: no Nomad Servers advertising service "nomad" in Consul datacenters: ["ap-southeast-1"]
    2018/10/18 06:10:08.926501 [WARN] manager: No servers available
    2018/10/18 06:10:13.636501 [ERR] http: Request /v1/agent/health?type=client, error: {"client":{"ok":false,"message":"no known servers"}}
    2018/10/18 06:10:23.637262 [ERR] http: Request /v1/agent/health?type=client, error: {"client":{"ok":false,"message":"no known servers"}}
    2018/10/18 06:10:25.905235 [WARN] manager: No servers available
    2018/10/18 06:10:25.906677 [ERR] client.consul: error discovering nomad servers: no Nomad Servers advertising service "nomad" in Consul datacenters: ["ap-southeast-1"]
    2018/10/18 06:10:33.638046 [ERR] http: Request /v1/agent/health?type=client, error: {"client":{"ok":false,"message":"no known servers"}}
    2018/10/18 06:10:43.640313 [ERR] http: Request /v1/agent/health?type=client, error: {"client":{"ok":false,"message":"no known servers"}}
    2018/10/18 06:10:45.124743 [WARN] manager: No servers available
    2018/10/18 06:10:45.126368 [ERR] client.consul: error discovering nomad servers: no Nomad Servers advertising service "nomad" in Consul datacenters: ["ap-southeast-1"]
    2018/10/18 06:10:53.641053 [ERR] http: Request /v1/agent/health?type=client, error: {"client":{"ok":false,"message":"no known servers"}}
    2018/10/18 06:11:03.641825 [ERR] http: Request /v1/agent/health?type=client, error: {"client":{"ok":false,"message":"no known servers"}}
    2018/10/18 06:11:05.791582 [WARN] manager: No servers available
    2018/10/18 06:11:08.487604 [WARN] manager: No servers available
    2018/10/18 06:11:08.489182 [ERR] client.consul: error discovering nomad servers: no Nomad Servers advertising service "nomad" in Consul datacenters: ["ap-southeast-1"]
    2018/10/18 06:11:13.642601 [ERR] http: Request /v1/agent/health?type=client, error: {"client":{"ok":false,"message":"no known servers"}}
```

**Comment by far-blue** [comment]
We're seeing the same thing. Reverted back to consul 1.2.3 and everything works.

**Comment by tantra35** [comment]
Just test this on test stand, and doens't confirm this, all works as expected

**Comment by lawliet89** [comment]
Let me know if more details are needed. There's nothing "unusual" about my setup. It was based off the Terrraform Nomad and Consul modules in this Github organization.

**Comment by rallister** [comment]
Got same issue on Ubuntu 18.04

**Comment by cgbaker** [comment]
Hi all, I have not been able to reproduce this on macOS, Ubuntu 16.04 or Ubuntu 18.04. 
If you could, please paste full server and client config, as well as server, client and consul logs.

**Comment by cgbaker** [comment]
Please re-open this issue with the configs necessary to reproduce this.

**Comment by alexdulin** [comment]
Managed to reproduce this issue using a somewhat crude example below. The problem seems to occur when consul servers are running a version lower than 1.3.0 (in this example 1.2.3), and the consul clients are running 1.3.0.

```bash
## Consul Server, running 1.2.3
mkdir -p /tmp/consul-server/data &&
cd /tmp/consul-server &&
wget --quiet -O ./consul.zip https://releases.hashicorp.com/consul/1.2.3/consul_1.2.3_linux_amd64.zip &&
unzip -q -o ./consul.zip &&
chmod +x ./consul &&
./consul agent \
  -server \
  -bind 127.0.0.1 \
  -data-dir /tmp/consul-server/data \
  -bootstrap-expect 1

## Consul Client, running 1.3.0
mkdir -p /tmp/consul-client/data &&
cd /tmp/consul-client &&
wget --quiet -O ./consul.zip https://releases.hashicorp.com/consul/1.3.0/consul_1.3.0_linux_amd64.zip &&
unzip -q -o ./consul.zip &&
chmod +x ./consul &&
consul agent \
  -client 127.0.0.1 \
  -bind 127.0.0.1 \
  -advertise 127.0.0.1 \
  -data-dir /tmp/consul-client/data \
  -retry-join=127.0.0.1:8301 \
  -server-port=8301 \
  -dns-port=8601 \
  -http-port=8501 \
  -serf-lan-port=8401 \
  -serf-wan-port=8402 \
  -node foobar



## Nomad server, running 0.8.6
mkdir -p /tmp/nomad-server/data &&
cd /tmp/nomad-server &&
wget --quiet -O ./nomad.zip https://releases.hashicorp.com/nomad/0.8.6/nomad_0.8.6_linux_amd64.zip &&
unzip -q -o ./nomad.zip &&
chmod +x ./nomad &&
cat <<EOM > server.hcl
consul {
  address             = "127.0.0.1:8501"
  server_service_name = "nomad"
  client_service_name = "nomad-client"
  auto_advertise      = true
  server_auto_join    = true
  client_auto_join    = true
}

server {
  enabled              = true
  bootstrap_expect     = 1
  raft_protocol        = 3
}

bind_addr = "0.0.0.0"

data_dir = "/tmp/nomad-server/data"

ports {
  http = 4646
  rpc  = 4647
  serf = 4648
}
EOM

nomad agent -server -config server.hcl


## Nomad client, running 0.8.6
mkdir -p /tmp/nomad-client/data &&
cd /tmp/nomad-client &&
wget --quiet -O ./nomad.zip https://releases.hashicorp.com/nomad/0.8.6/nomad_0.8.6_linux_amd64.zip &&
unzip -q -o ./nomad.zip &&
chmod +x ./nomad &&
cat <<EOM > client.hcl
consul {
  address             = "127.0.0.1:8501"
  server_service_name = "nomad"
  client_service_name = "nomad-client"
  auto_advertise      = true
  server_auto_join    = true
  client_auto_join    = true
}

bind_addr = "0.0.0.0"

data_dir = "/tmp/nomad-client/data"

ports {
  http = 4746
  rpc  = 4747
  serf = 4748
}
EOM

nomad agent -client -config client.hcl
```

**Comment by cgbaker** [comment]
Okay, reproduced this... will re-open and put updates here. thanks again, @alexdulin

**Comment by rallister** [comment]
Hi,
All my consul & servers are 1.3. Initially client could not see the server in the catalog so I had to add server_join directive in hcl config.
`server_join {
    retry_join = ["provider=aws ...
  }
`
I've noticed that in your config you're specifying the service name of nomad server and was going to try that as well:
`
server_service_name = "nomad"
`
But after I've commented out server_join clause and restarted nomad server was found straight away. So it seems like the issue went away for me without specifying server_service_name. 
Since I don't know why did it start to work I'll leave the server_join clause in place for now.
Thanks.

**Comment by cgbaker** [comment]
Okay, I've found the underlying issue. The endpoint that the Nomad server is querying on the Consul client is returning an invalid response. Specifically, the call 
`GET /v1/catalog/service/nomad?...&tag=rpc`
is returning an empty list `[]`.

The cause of this is the version mismatch between the Consul client and the Consul server; it is not really supported to mix different "major" versions of Consul. The issue, as noted above, is resolved when the Consul upgrade is completed.

I've pointed this out to the Consul team and they are going to add a note to their upgrade guide.

Thanks all for your help in tracking this one down.

**Comment by alexdulin** [comment]
Thanks @cgbaker, you are definitely correct. After following the Consul upgrade guides and updating the consul binary version on the server agents first, then updating clients, I can confirm that the issue does not occur (have not seen it once while updating 80+ nodes).

**Comment by cgbaker** [comment]
By the way, this issue was just filed in Consul and fixed in a 1.3.1 release: 
https://github.com/hashicorp/consul/blob/v1.3.1/CHANGELOG.md
